src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -266,31 +266,72 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
         val accu1 = if (accu exists (_ derivesFrom c)) accu else c :: accu
         if (cs == c.baseClasses) accu1 else dominators(rest, accu1)
     }
-    if (ctx.featureEnabled(defn.LanguageModuleClass, nme.keepUnions)) tp
-    else tp match {
-      case tp: OrType =>
-        def isClassRef(tp: Type): Boolean = tp match {
-          case tp: TypeRef => tp.symbol.isClass
-          case tp: RefinedType => isClassRef(tp.parent)
-          case _ => false
-        }
-        def next(tp: TypeProxy) = tp.underlying match {
-          case TypeBounds(_, hi) => hi
-          case nx => nx
-        }
-        tp.tp1 match {
-          case tp1: TypeProxy if !isClassRef(tp1) =>
-            approximateUnion(next(tp1) | tp.tp2)
-          case _ =>
-            tp.tp2 match {
-              case tp2: TypeProxy if !isClassRef(tp2) =>
-                approximateUnion(tp.tp1 | next(tp2))
+    def approximateOr(tp1: Type, tp2: Type)(implicit ctx: Context): Type = {
+      def isClassRef(tp: Type): Boolean = tp match {
+        case tp: TypeRef => tp.symbol.isClass
+        case tp: RefinedType => isClassRef(tp.parent)
+        case _ => false
+      }
+      def next(tp: TypeProxy) = tp.underlying match {
+        case TypeBounds(_, hi) => hi
+        case nx => nx
+      }
+      /** If `tp1` and `tp2` are typebounds, try to make one fit into the other
+       *  or to make them equal, by instantiating uninstantiated type variables.
+       */
+      def homogenizedUnion(tp1: Type, tp2: Type): Type = {
+        tp1 match {
+          case tp1: TypeBounds =>
+            tp2 match {
+              case tp2: TypeBounds =>
+                def fitInto(tp1: TypeBounds, tp2: TypeBounds): Unit = {
+                  val nestedCtx = ctx.fresh.setNewTyperState
+                  if (tp2.boundsInterval.contains(tp1.boundsInterval)(nestedCtx))
+                    nestedCtx.typerState.commit()
+                }
+                fitInto(tp1, tp2)
+                fitInto(tp2, tp1)
               case _ =>
-                val commonBaseClasses = tp.mapReduceOr(_.baseClasses)(intersect)
-                val doms = dominators(commonBaseClasses, Nil)
-                doms.map(tp.baseTypeWithArgs).reduceLeft(AndType.apply)
             }
+          case _ =>
         }
+        tp1 | tp2
+      }
+
+      tp1 match {
+        case tp1: RefinedType =>
+          tp2 match {
+            case tp2: RefinedType if tp1.refinedName == tp2.refinedName =>
+              return tp1.derivedRefinedType(
+                approximateUnion(OrType(tp1.parent, tp2.parent)),
+                tp1.refinedName,
+                homogenizedUnion(tp1.refinedInfo, tp2.refinedInfo).substRefinedThis(tp2, RefinedThis(tp1)))
```

Comment:
Why not `homogenizedUnion(tp1.refinedInfo, tp2.refinedInfo.substRefinedThis(tp2, RefinedThis(tp1)))` ? This should require less traversals and is closer to what we were doing before https://github.com/lampepfl/dotty/commit/ee76fda79d446a2d6db51cb4af032a8e92936013